Wrappers/Python/cil/optimisation/functions/KullbackLeibler.py

```diff
@@ -226,281 +436,145 @@ def force_jit():
-                out.as_array()[tmp_sum_array>0] = \
-                    1 - self.b.as_array()[tmp_sum_array>0]/tmp_sum_array[tmp_sum_array>0]
+        """           
 
-            
-    def convex_conjugate(self, x):
-        
-        r"""Returns the value of the convex conjugate of the KullbackLeibler function at :math:`(b, x + \eta)`.                
-        
-        .. math:: F^{*}(b, x + \eta) = - b \log(1-x^{*}) - <x^{*}, \eta> 
-        
-        """  
-        if self.use_numba and has_numba:
+        if out is None:
+            out = (x * 0.)
+            out_np = out.as_array()
             if self.mask is not None:
-                return kl_convex_conjugate_mask(x.as_array(), self.b_np, self.eta_np, self.mask)
-            return kl_convex_conjugate(x.as_array(), self.b_np, self.eta_np)
+                kl_gradient_mask(x.as_array(), self.b.as_array(), out_np, self.eta.as_array(), self.mask)
+            kl_gradient(x.as_array(), self.b.as_array(), out_np, self.eta.as_array())
+            out.fill(out_np)
+            return out
         else:
-            tmp = 1 - x.as_array()
-            ind = tmp>0
-            xlogy = - scipy.special.xlogy(self.b.as_array()[ind], tmp[ind])  
-            return numpy.sum(xlogy) - self.eta.dot(x)
-            
-    def proximal(self, x, tau, out=None):
-        
+            out_np = out.as_array()
+            kl_gradient(x.as_array(), self.b.as_array(), out_np, self.eta.as_array())
+            out.fill(out_np)
+
+    def proximal(self, x, tau, out = None):
+
         r"""Returns the value of the proximal operator of the KullbackLeibler function at :math:`(b, x + \eta)`.
         
-        .. math:: \mathrm{prox}_{\tau F}(x) = \frac{1}{2}\bigg( (x - \eta - \tau) + \sqrt{ (x + \eta - \tau)^2 + 4\tau b} \bigg)
-        
-        The proximal for the convex conjugate of :math:`F` is 
-        
-        .. math:: \mathrm{prox}_{\tau F^{*}}(x) = 0.5*((z + 1) - \sqrt{(z-1)^2 + 4 * \tau b})
+        :math:`\mathrm{prox}_{\tau F}(x) = \frac{1}{2}\bigg( (x - \eta - \tau) + \sqrt{ (x + \eta - \tau)^2 + 4\tau b} \bigg)`
+                            
+        """        
         
-        where :math:`z = x + \tau \eta`
-                    
-        """
-        if self.use_numba and has_numba:
-            if out is None:
-                out = (x * 0.)
-                # out_np = numpy.empty_like(out.as_array(), dtype=numpy.float64)
-                out_np = out.as_array()
-                if isinstance(tau, Number):
-                    if self.mask is not None:
-                        kl_proximal_mask(x.as_array(), self.b_np, tau, out_np, \
-                            self.eta_np, self.mask)
-                    else:
-                        kl_proximal(x.as_array(), self.b_np, tau, out_np, self.eta_np)
+        if out is None:
+            out = (x * 0.)
+            # out_np = numpy.empty_like(out.as_array(), dtype=numpy.float64)
+            out_np = out.as_array()
```

Comment:
I think this initial `if` should check if the `out` exists or not and then create `out` and `out_np` then, the rest of the method would perfectly work, except that you need to remember to return in the case of `out` was passed as None.